sphinx/ext/napoleon/docstring.py

```diff
@@ -872,6 +884,83 @@ def __init__(self, docstring: Union[str, List[str]], config: SphinxConfig = None
         self._directive_sections = ['.. index::']
         super().__init__(docstring, config, app, what, name, obj, options)
 
+    def _convert_type_spec(self, _type):
+        def recombine_set(tokens):
+            def combine_set(tokens):
+                in_set = False
+                set_items = []
+
+                for token in tokens:
+                    if token.startswith("{"):
+                        in_set = True
+                    elif token.endswith("}"):
+                        in_set = False
+                        set_items.append(token)
+
+                    if in_set:
+                        set_items.append(token)
+                    else:
+                        if set_items:
+                            token = "".join(set_items)
+                            set_items = []
+                        yield token
+
+            return list(combine_set(tokens))
+
+        def tokenize_type_spec(spec):
+            delimiters = r"(\sor\s|\sof\s|:\s|,\s|[{]|[}])"
+
+            split = [
+                item
+                for item in re.split(delimiters, _type)
+                if item
+            ]
+            tokens = recombine_set(split)
+            return tokens
+
+        def token_type(token):
+            if token.startswith(" ") or token.endswith(" "):
+                type_ = "delimiter"
+            elif token.startswith("{") and token.endswith("}"):
+                type_ = "value_set"
+            elif token in ("optional", "default"):
+                type_ = "control"
+            elif re.match(":[^:]+:`[^`]+`", token):
+                type_ = "reference"
+            elif token.isnumeric() or (token.startswith('"') and token.endswith('"')):
+                type_ = "literal"
+            else:
+                type_ = "obj"
+
+            return type_
+
+        def convert_obj(obj, translations):
+            return translations.get(obj, ":obj:`{}`".format(obj))
+
+        tokens = tokenize_type_spec(_type)
+        types = [
+            (token, token_type(token))
+            for token in tokens
+        ]
+
+        # TODO: make this configurable
+        translations = {
+            "sequence": ":term:`sequence`",
+            "mapping": ":term:`mapping`",
+            "dict-like": ":term:`dict-like <mapping>`",
+        }
+
+        converters = {
+            "value_set": lambda x: f":noref:`{x}`",
+            "literal": lambda x: f":noref:`{x}`",
+            "obj": lambda x: convert_obj(x, translations),
+            "control": lambda x: f":noref:`{x}`",
```

Comment:
I still don't understand why `:noref:` role is needed. What is different from `lambda x: x`?

Reply:
without this PR,
```
param1 : str, optional
```
gets translated to
```
:param: param1
:type: str, optional
```
and sphinx will try to reference both `str` and `optional` (using `:obj:`?). To make sure `optional` (or any literal / modifier, etc.) does not get referenced, we could instead use
```
:param: param1
:type: :obj:`str`, optional
```
but we can't do that if there's no real type (which is valid):
```
:param: param1
:type: optional
```
In that case, the only way to still make sure `optional` does not get treated as a type I could find is by explicitly marking it with a no-op role, hence the `:noref:`.

Reply:
Please wait a moment. I consider the way not to use the role. I think it's not good to add `:noref:` role globally because it is not semantic...

Note: We can stop the behavior, text emphasis can be used (ex. `:type: *optional*`). The python domain check the type is single text or not. So the marked up text (like `:noref:` and emphasis) is not converted to the reference.

Reply:
as I said the new role was the only way I could find, but I didn't try text emphasis.

Since text emphasis changes the way the text is displayed, we need to consider how to display the different token types: Should control words (`optional`, `default`) be displayed as *optional* or **optional**? What about literals (strings, numbers, booleans, value sets)?

I'd argue we should try to use different styles for both to make the difference visible. To do that, I'd like to propose using double backticks for literals (`True`, `False` and `None` can be referenced):
```
``"xyz"``
``{"abc", "def"}``
``1.4``
``10``
:obj:`True`
:obj:`None`
```
and emphasis for control tokens (`*optional*` and `*default*`). I don't feel strongly about this, though.

Also, the role I chose as default for type tokens (`:obj:`) is displayed differently from the default role of the `:type:` field. I think we should try to be consistent here (either always use `:obj:` or always use the default role of the `:type:` field).